Tests failing on Python 2

There are two test cases that random fails on Python 2.7.

```python
========================= FAILURES ==========================
_________________ test_search_can_paginate __________________

client = <tests.conftest.FastRetryClient object at 0x7fb1bf4b8890>, requests_mocker = <requests_mock.mocker.Mocker object at 0x7fb1bf50d290>

    def test_search_can_paginate(client, requests_mocker):
        """Paginated search will stop issuing requests if pages are no longer referenced."""
    
        client._PAGE_SIZE = 10
    
        expected_repos = []
        responses = []
        current_response = []
        for i in range(0, 997):
            repo_id = "repo-%s" % i
            expected_repos.append(Repository(id=repo_id))
            current_response.append({"id": repo_id})
            if len(current_response) == client._PAGE_SIZE:
                responses.append({"json": current_response})
                current_response = []
    
        responses.append({"json": current_response})
    
        requests_mocker.post(
            "https://pulp.example.com/pulp/api/v2/repositories/search/", responses
        )
    
        # Imagine we have some function which processes two pages of results and
        # then stops
        def do_something_first_two_pages(repos_f):
            page = repos_f.result()
            return len(page.data) + len(page.next.result().data)
    
        two_pages_len = do_something_first_two_pages(client.search_repository())
    
        # It should have seen the first two pages
        assert two_pages_len == 20
    
        # Now wait a bit to ensure any futures under the hood have time to finish up
        time.sleep(2.0)
    
        # It should have done only two or three requests
        # (2 vs 3 depends on scheduling). Because we threw away all references to
        # the search result, it stopped searching rather than continuing through
        # all pages.
>       assert requests_mocker.call_count in (2, 3)
E       assert 43 in (2, 3)
E        +  where 43 = <requests_mock.mocker.Mocker object at 0x7fb1bf50d290>.call_count

tests/client/test_search_stops.py:45: AssertionError
____________________ test_publish_cancel ____________________

fast_poller = None, requests_mocker = <requests_mock.mocker.Mocker object at 0x7fb1bfff13d0>, client = <tests.conftest.FastRetryClient object at 0x7fb1bf65c550>
caplog = <_pytest.logging.LogCaptureFixture object at 0x7fb1bc6bc790>

    def test_publish_cancel(fast_poller, requests_mocker, client, caplog):
        """Cancelling a publish future will cancel running Pulp task(s)."""
    
        repo = YumRepository(
            id="some-repo",
            distributors=(
                Distributor(id="yum_distributor", type_id="yum_distributor"),
                Distributor(id="cdn_distributor", type_id="cdn_distributor"),
            ),
        )
        repo.__dict__["_client"] = client
    
        requests_mocker.post(
            "https://pulp.example.com/pulp/api/v2/repositories/some-repo/actions/publish/",
            [{"json": {"spawned_tasks": [{"task_id": "task1"}]}}],
        )
    
        task_search_url = "https://pulp.example.com/pulp/api/v2/tasks/search/"
        requests_mocker.post(
            task_search_url, [{"json": [{"task_id": "task1", "state": "running"}]}]
        )
    
        requests_mocker.delete("https://pulp.example.com/pulp/api/v2/tasks/task1/")
    
        # Start the publish
        publish_f = repo.publish()
    
        # Wait until we're sure poll thread has seen this task
        for _ in range(0, 1000):
            if (
                requests_mocker.last_request
                and requests_mocker.last_request.url == task_search_url
            ):
                break
            time.sleep(0.001)
        assert requests_mocker.last_request.url == task_search_url
    
        # We should be able to cancel it
        assert publish_f.cancel()
    
        # It should have cancelled the underlying Pulp task
        req = requests_mocker.last_request
>       assert req.url == "https://pulp.example.com/pulp/api/v2/tasks/task1/"
E       AssertionError: assert 'https://pulp...tasks/search/' == 'https://pulp..../tasks/task1/'
E         - https://pulp.example.com/pulp/api/v2/tasks/search/
E         ?                                             ^^^^^
E         + https://pulp.example.com/pulp/api/v2/tasks/task1/
E         ?   
```

These tests do not fail every single time, they usually fail if you run twice or more.

I'm using `tox -e py27` to run tests, the same way that Travis CI does and it also fails on Travis as shown in PR #20.

--- a/tests/client/test_search_stops.py
+++ b/tests/client/test_search_stops.py
@@ -1,10 +1,11 @@
 import time
-from pubtools.pulplib import Repository
+from pubtools.pulplib import Client
 
 
-def test_search_stops_paginate(client, requests_mocker):
+def test_search_stops_paginate(requests_mocker):
     """Paginated search will stop issuing requests if pages are no longer referenced."""
 
+    client = Client("https://pulp.example.com/")
     client._PAGE_SIZE = 10
 
     responses = []
@@ -17,6 +18,21 @@ def test_search_stops_paginate(client, requests_mocker):
             current_response = []
 
     responses.append({"json": current_response})
+
+    # Inject a failing response for stable behavior from test:
+    #
+    # The expected behavior is that, after the first two requests,
+    # pagination is cancelled since we no longer hold onto a page.
+    # But that cancellation is racing with the scheduling of the
+    # request, and there is generally no way to ensure our cancel
+    # happens before the request thread starts the request, so we
+    # can't be sure about how many requests we'd do.
+    #
+    # However, if we insert a failure after the first two pages,
+    # then the client will need to retry a request (with some delay).
+    # It should always be possible to cancel the request during that
+    # retry delay, making the test stable.
+    responses.insert(2, {"status_code": 500})
 
     requests_mocker.post(
         "https://pulp.example.com/pulp/api/v2/repositories/search/", responses

--- a/tests/client/test_search_stops.py
+++ b/tests/client/test_search_stops.py
@@ -7,12 +7,10 @@ def test_search_stops_paginate(client, requests_mocker):
 
     client._PAGE_SIZE = 10
 
-    expected_repos = []
     responses = []
     current_response = []
     for i in range(0, 997):
         repo_id = "repo-%s" % i
-        expected_repos.append(Repository(id=repo_id))
         current_response.append({"id": repo_id})
         if len(current_response) == client._PAGE_SIZE:
             responses.append({"json": current_response})

--- a/tests/client/test_search_stops.py
+++ b/tests/client/test_search_stops.py
@@ -2,7 +2,7 @@ import time
 from pubtools.pulplib import Repository
 
 
-def test_search_can_paginate(client, requests_mocker):
+def test_search_stops_paginate(client, requests_mocker):
     """Paginated search will stop issuing requests if pages are no longer referenced."""
 
     client._PAGE_SIZE = 10

--- a/tests/repository/test_publish_cancel.py
+++ b/tests/repository/test_publish_cancel.py
@@ -44,6 +44,10 @@ def test_publish_cancel(fast_poller, requests_mocker, client, caplog):
     assert publish_f.cancel()
 
     # It should have cancelled the underlying Pulp task
-    req = requests_mocker.last_request
-    assert req.url == "https://pulp.example.com/pulp/api/v2/tasks/task1/"
-    assert req.method == "DELETE"
+    task_req = [
+        r
+        for r in requests_mocker.request_history
+        if r.url == "https://pulp.example.com/pulp/api/v2/tasks/task1/"
+    ]
+    assert len(task_req) == 1
+    assert task_req[0].method == "DELETE"

